session/session.go

```diff
@@ -11,447 +11,243 @@
+		var errMsg ws.Error
+		msgpack.Unmarshal(msg.Body, &errMsg) //nolint:errcheck
+		log.Errorf("session: received error from client: %s", errMsg.Error)
+		close = errMsg.Close
 
-func (s *MenderShellSession) GetShellCommandPath() string {
-	return s.command.Path
-}
-
-func (s *MenderShellSession) StartShell(sessionId string, terminal MenderShellTerminalSettings) error {
-	if s.status == ActiveSession || s.status == HangedSession {
-		return ErrSessionShellAlreadyRunning
+	default:
+		sess.Error(msg, false, fmt.Sprintf(
+			"session: control type message not understood: '%s'",
+			msg.Header.MsgType,
+		))
 	}
-
-	pid, pseudoTTY, cmd, err := shell.ExecuteShell(
-		terminal.Uid,
-		terminal.Gid,
-		terminal.HomeDir,
-		terminal.Shell,
-		terminal.TerminalString,
-		terminal.Height,
-		terminal.Width)
-	if err != nil {
-		return err
-	}
-
-	//MenderShell represents a process of passing messages between backend
-	//and the shell subprocess (started above via shell.ExecuteShell) over
-	//the websocket connection
-	log.Infof("mender-connect starting shell command passing process, pid: %d", pid)
-	s.shell = shell.NewMenderShell(sessionId, pseudoTTY, pseudoTTY)
-	s.shell.Start()
-
-	s.shellPid = pid
-	s.reader = pseudoTTY
-	s.writer = pseudoTTY
-	s.status = ActiveSession
-	s.terminal = terminal
-	s.pseudoTTY = pseudoTTY
-	s.command = cmd
-	s.activeAt = timeNow()
-
-	// start the healthcheck go-routine
-	go s.healthcheck()
-
-	return nil
-}
-
-func (s *MenderShellSession) GetId() string {
-	return s.id
-}
-
-func (s *MenderShellSession) GetShellPid() int {
-	return s.shellPid
+	return
 }
 
-func (s *MenderShellSession) IsExpired(setStatus bool) bool {
-	if defaultSessionIdleExpiredTimeout != NoExpirationTimeout {
-		idleTimeoutReached := s.activeAt.Add(defaultSessionIdleExpiredTimeout)
-		return timeNow().After(idleTimeoutReached)
+func (sess *Session) Ping() error {
+	ping := &ws.ProtoMsg{
+		Header: ws.ProtoHdr{
+			Proto:     ws.ProtoTypeControl,
+			MsgType:   ws.MessageTypePing,
+			SessionID: sess.ID,
+		},
 	}
-	e := timeNow().After(s.expiresAt)
-	if e && setStatus {
-		s.status = ExpiredSession
+	return sess.w.WriteProtoMsg(ping)
+}
+
+func funcname(fn string) string {
+	// strip package path
+	i := strings.LastIndex(fn, "/")
+	fn = fn[i+1:]
+	// strip package name.
+	i = strings.Index(fn, ".")
+	fn = fn[i+1:]
+	return fn
+}
+
+func (sess *Session) handlePanic() {
+	if r := recover(); r != nil {
+		var stacktrace strings.Builder
+		var trace [MaxTraceback]uintptr
+		num := runtime.Callers(3, trace[:])
+		for i := 0; i < num; i++ {
+			fn := runtime.FuncForPC(trace[i])
+			if fn == nil {
+				fmt.Fprintf(&stacktrace, "\n???")
+				continue
+			}
+			file, line := fn.FileLine(trace[i])
+			fmt.Fprintf(&stacktrace, "\n%s:%d.%s",
+				file, line, funcname(fn.Name()),
+			)
+		}
+		log.WithField("trace", stacktrace.String()).
+			Errorf("[panic] %s", r)
+		sess.Error(&ws.ProtoMsg{}, true, "internal error")
+	}
+	close(sess.done)
+}
```

Comment:
this is slightly too much for me. why are we doing that?

Reply:
This was mostly taken from the `accesslog` middleware. I don't we should terminate on panic in handler functions - simply closing and freeing the session should be enough. The stacktrace logger field is simply for debugging purposes.